Deduplication to avoid infinite loops. Better priority queue for multiple domains.

ivanprado opened the pull request:
A new dedup filter was included which allows having different deduplication sets. This is handy to have independent deduplicaiton for  AutoExtract and Scrapy requests. This approach avoids deduplication cycles (e.g. a site with links to www.examplesite.com/contact that are redirected to www.examplesite.com/contact/)

Also scrapy.pqueues.DownloaderAwarePriorityQueue was enabled for better scheduling and concurrency

I removed `dont_filter=True`, from here:
https://github.com/scrapinghub/autoextract-spiders/blob/master/autoextract_spiders/spiders/crawler_spider.py#L267
and here:
https://github.com/scrapinghub/autoextract-spiders/blob/master/autoextract_spiders/spiders/autoextract_spider.py#L170 . I think is the right thing to do because instead, we have individual deduplication to AutoExtract and Scrapy requests. But I wonder if I'm missing something and `dont_filter=True` was required for some particular reason. Reviewers, what do you think? Is maybe related with Frontera somehow?

I also found other places with `dont_filter=True`, for example here where the feeds are loaded: https://github.com/scrapinghub/autoextract-spiders/blob/master/autoextract_spiders/spiders/autoextract_article.py#L64

I also think that they should be removed, but I'm unsure. Do you know why are they required?

ivanprado committed:
Deduplication to avoid cycles. Better priority queue for multiple domains.

A new dedup filter was included which allows to have different sets of
deduplication. This is handy to have a set for AutoExtract requests
and a different one for Scrapy ones.

Also scrapy.pqueues.DownloaderAwarePriorityQueue was enabled for better
scheduling and concurrency

vshlapakov commented on autoextract_spiders/settings.py:
```diff
@@ -8,6 +8,7 @@
 #     http://doc.scrapy.org/en/latest/topics/settings.html
 #     http://scrapy.readthedocs.org/en/latest/topics/downloader-middleware.html
 #     http://scrapy.readthedocs.org/en/latest/topics/spider-middleware.html
+import autoextract_spiders
```
Is there any implicit reason to add it here?

ivanprado replied:
Ops. Removed here: https://github.com/scrapinghub/autoextract-spiders/pull/25/commits/0ed04ba41c109231a64f448d02579137421936bc

vshlapakov commented on autoextract_spiders/spiders/autoextract_spider.py:
```diff
@@ -162,12 +163,12 @@ def make_extract_request(self, url, meta=None, check_page_type=True):
                                  page_type=self.page_type,
                                  callback=self.parse_item,
-                                 errback=self.errback_item,
-                                 dont_filter=True)
```
Not sure we should drop it tbh, see the review comment

vshlapakov reviewed (commented):
The only reason behind `dont_filter` is for sure forcing specific requests to bypass the duplication filter. While it makes sense to have it in some specific places (for example, we might want to retry some requests to AutoExtract API on bans/timeouts), I think still the main reason here is allowing to bypass Frontera/HCF deduplication when running from scratch with the same seeds, otherwise the requests might be dropped as already visited.

ivanprado committed:
Remove unused import

ivanprado commented:
@vshlapakov thank you for the feedback, is being very useful. 

> for example, we might want to retry some requests to AutoExtract API on bans/timeouts

Yes, we should improve this logic. The only retrial I see now is for HTTP status 429 (too many requests) but I wonder if this is really working, as I think that `AutoExtractMiddleware` is generating an exception for it so the response never arrives at the Retry middleware. What do you think about that?

> Frontera/HCF

I'm really lost here. Never used Frontera/HCF by the time. Could you help me understand that? Why is required Frontera/HCF integration in autoextract-spider? I mean, which is the use case, what is it solving here.

vshlapakov commented:
> The only retrial I see now is for HTTP status 429

Yeah, the case about retrying is more a speculation on my side, I'd ask @croqaz to elaborate on this, I have only the case with Frontera/HCF in my head.

> Never used Frontera/HCF by the time. Could you help me understand that?

The key benefit of using Frontera here is providing resumable crawls, so you could stop it and continue anytime later. It also provides scaling capabilities out of the box, so it's possible to run multiple discovery-producers/worker-consumers in parallel (not sure if we need it in the short terms though).  It makes it work by using a custom Scrapy scheduler & a few middlewares, so the requests are stored not locally but via [HCF](https://doc.scrapinghub.com/api/frontier.html) backed by Hubstorage.

A few links to bring more light into it:
- https://frontera.readthedocs.io/en/latest/topics/what-is-cf.html
- https://github.com/scrapinghub/shub-workflow/wiki/Basic-Tutorial (we don't use shub-workflow here directly, but the concepts are the same)
- https://github.com/scrapinghub/autoextract-spiders/pull/1 (the integration changes PR)

The technique is well-tested in many production projects, it's not that difficult to use it despite it does require some efforts to familiarize with the concepts.

ivanprado commented:
Thank you very much @vshlapakov for the links, now I'm starting to understand better. Another question, I see that `cf_store=True` is only set for AutoExtract request, here:

https://github.com/scrapinghub/autoextract-spiders/blob/master/autoextract_spiders/spiders/autoextract_spider.py#L164

Does that mean that only AutoExtract requests are saved in the frontier? Or are other parts/request that are also managed by the frontier in this code?

croqaz commented:
> Does that mean that only AutoExtract requests are saved in the frontier? Or are other parts/request that are also managed by the frontier in this code?

These spiders make discover requests, and AutoExtract requests too. Ideally only the discovery requests are saved in CF. That's where we get the benefit to resume the crawl, by ignoring the links that were already crawled.

ivanprado commented:
> Ideally only the discovery requests are saved in CF

Thank @croqaz. I think that the ones that are saved are AutoExtract requests, but that's Ok because they are symmetric. One AE request per discovery request.

croqaz commented:
I think the `dont_filter=True` are only supposed to be there in the initial requests, made directly from the seeds, and also when the links to articles/ products/ whatever vertical, are provided directly by the user.
Also I can see it's there when parsing the RSS/ Atom feeds, because we want to scrape them everytime.
I think because the requests are eventually saved and remembered by HCF, if we don't provide the `dont_filter=True`, they will be dropped, in case they were crawled before (but I haven't actually tested this now)

ivanprado committed:
BFO crawling and referrer_url

ivanprado committed:
referrer_url missing in one place

ivanprado committed:
Force crawling of the seeds

ivanprado commented:
Ey @croqaz and @vshlapakov. I have been using these changes with success in my experiments, so I feel more confident they are right. I propose to merge them in autoextract-spiders as they are useful. What do you think?

ivanprado committed:
remove referrer_url

vshlapakov commented:
@ivanprado Sorry, just got back from a short vacation. I do think the changes are useful overall, but I don't get your point/argument on dropping the `dont_filter` parameters, as your tests don't cover using Frontera/HCF for resumable crawls as I mentioned on top of the thread

> the main reason here is allowing to bypass Frontera/HCF deduplication when running from scratch with the same seeds, otherwise the requests might be dropped as already visited.

Changing the logic would mean ignoring seeds/initial requests even if we want to recrawl it, and I'd like to be on the same page regarding a whole picture before modifying the current behavior. If you insist on dropping it, let's continue the discussion, otherwise I'd be happy to merge the rest of the changes.

ivanprado commented:
Thanks @vshlapakov, makes sense :+1: . Let's resume again the `dont_filter` discussion. To summarize, I found two problems in existing code:

1. The same item found in different pages is extracted again an again as it is always passing the dedup filter, generating duplicated items and wasting resources
2. Infinite loops can appear. See the following example:

> www.site.com/product1 contains a link to www.site.com/product1/ and the site redirects requests from www.site.com/product1/ to www.site.com/product1

So in the existing code the following happens:

1. AutoExtract request to www.site.com/product1/
2. Gets redirected to www.site.com/product1 and the product is extracted
3. Scrapy request to www.site.com/product1 to follow all links. A link to www.site.com/product1/ is found and followed without filtering, so loop to point 1.

So my solution proposed in this PR is to have independent local deduplication queues, one for AutoExtract requests (fixes problem 1.), and other for the Scrapy requests (fixes problem 2.).

But according to your comments, it seems this won't work well with Frontera/HCF and resumable crawls. And here is where I'm weaker, so let me know if the following reasoning is right or not.

I guess that each time you resume crawling, the Scrapy deduplication filters are empty (because it is a new run), so the unique filtering working will be the one from Frontera/HCF, that is applied only to AutoExtract requests.

The problem it seems to be if you want to force some seeds to be fetched again. Having `dont_filter=True` will do the trick, but won't deduplicate AutoExtract requests in any case, so problem 1. will be still present. It doesn't look like a solution for me, but probably I'm missing something. `dont_filter=True` seems to me like completely ignore Frontera/HCF deduplication and use it only to know where to continue in next run. But this will potentially end in a completely fully re-crawl of the site on each run. 

What do you think, is this reasoning right or am I missing anything?

vshlapakov commented:
Nice @ivanprado, thanks a lot for explaining your position, I understand it better now :+1: Let's see if I can answer some of your concerns or not.

>  this will potentially end in a completely fully re-crawl of the site on each run

Yeah, that's correct and expected in some cases, we do want to be able to re-crawl from scratch. 

However your point about duplicated items seems solid enough, I haven't seen such cases but it might happen. Maybe we could make the dont-filter optional and enable it with a param when it's appropriate?

kmike committed:
Merge pull request #26 from scrapinghub/bfo

Breadth-first order for crawling

ivanprado commented:
> we do want to be able to re-crawl from scratch

@vshlapakov in such a case wouldn't be better to remove the data from the Frontera/HCF using `hcfpal.py` as mentioned in the documentation (https://github.com/scrapinghub/autoextract-spiders#url-deduplication)? To me, it seems a better solution that just disabling filters, especially because it would generate duplicates of the same item very likely (one each time a link for the item is found in the site).

vshlapakov commented:
>  in such a case wouldn't be better to remove the data from the Frontera/HCF

@ivanprado  That's a good point, I believe we could go this way when recrawling from scratch, and that would work better as a whole system :+1: What about explicitly revisiting seeds w/o removing all data for discovery purpose? There's a very concrete use-case with some feed URLs updated from time to time. Maybe we could have a flag to revisit seeds that would bypass the filter provided by HCF, but taking the local deduplication queue into account, wdyt? I feel like we'll need to work on the description of the most common use-cases which could be handled by the project, it'd be super helpful in discussions like this one.

ivanprado committed:
Directly extract links from rss instead of send them to discovery

Otherwise:
- The link in the rss is never extracted itself
- Starts discovery when would not be desired
- Don't allow effective dont_filter setup

ivanprado committed:
Force dont_filter for single url (testing pourposes), and seeds.

ivanprado commented:
> What about explicitly revisiting seeds w/o removing all data for discovery purpose?

 @vshlapakov from my point of view it only makes sense for direct AutoExtract extractions, not for discovery. In other words, I see Ok to:
- Force AutoExtract extraction of all URLs in feed (but avoiding force discovery)
- Force AutoExtract extraction of links in rss feeds found in seeds
- Force extraction on links found in seeds (but not going further)
- Always have the deduplication enabled for discovery. No possibility to override apart of the listed cases above.

Note that this is the actual behaviour in this PR (after this new commit https://github.com/scrapinghub/autoextract-spiders/pull/25/commits/56b4f67563cb9858c4a0d3691a6f9239deeb7fe1) 

I also found a bug in rss link extraction. I fixed it in https://github.com/scrapinghub/autoextract-spiders/pull/25/commits/5e813c3bd54f2de5251fff8d5c352bf131f7650f

vshlapakov commented:
@ivanprado Thanks! Tbh I'm still not sure I completely follow the logic. Could you please describe shortly how it's going to work with a simple case? Say we have a seed URL which is a link to RSS feed, updated every day, old entries are getting replaced by the new one gradually. Would it be possible to run the spider repeatably and get the new links? 

Asking because I have a feeling that in this case the seed link would be stored in Frontera/HCF slot and I don't see a way how to force to ignore it aside for deleting all data from there, so failing back to the local deduplication only (which won't work b/w the crawls). My main worry here is that we introduce/fix the local deduplication (which is good and useful, no doubt) but break some things in the remote deduplication, so my intention here is get on the same page about evolving it in the future. Let's maybe have a call early next week and discuss the details, it could save us some time.

vshlapakov reviewed (approved):
@ivanprado @croqaz Alright, I don't want to delay the progress even more, let's build it on the real case scenarios rather than artificial ones, step by step.  I agree the changes are good enough to be merged, let's move forward and address any other related issues later as they occur. Thanks a ton for providing all requested details/feedback and for the great discussion overall :+1:

croqaz reviewed (approved):
👍

ivanprado commented:
Thanks @croqaz and @vshlapakov for the reviews. 

And thank you @vshlapakov for the green light. I'm aware the solution is not yet perfect, but as you said, step by step, and I think this PR is a step in right direction.
